File: rust/src/code/diameter_of_binary_tree.rs
use std::cell::RefCell;
use std::rc::Rc;

use super::models::binary_tree_node::TreeNode;

struct Solution {}

impl Solution {
    pub fn diameter_of_binary_tree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut max: i32 = 0;
        Self::get_diameter(&root, &mut max);

        max
    }

    fn get_diameter(root: &Option<Rc<RefCell<TreeNode>>>, max: &mut i32) -> i32 {
        if let Some(node) = root {
            let left = Self::get_diameter(&node.borrow().left, max);
            let right = Self::get_diameter(&node.borrow().right, max);
            *max = std::cmp::max(left + right, *max);
            std::cmp::max(left, right) + 1
        } else {
            0
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;
    use crate::code::utils::deserialize_to_binary_tree::deserialize_to_BT;

    #[test]
    fn diameter_of_binary_tree_case_1() {
        // arrange
        let root = vec![Some(1), Some(2), Some(3), Some(4), Some(5)];
        let expected = 3;

        // act
        let actual = Solution::diameter_of_binary_tree(deserialize_to_BT(root));

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn diameter_of_binary_tree_case_2() {
        // arrange
        let root = vec![Some(1), Some(2)];
        let expected = 1;

        // act
        let actual = Solution::diameter_of_binary_tree(deserialize_to_BT(root));

        // assert
        assert_eq!(expected, actual);
    }
}

File: python/diameter_of_binary_tree.py
import sys

sys.path.append("./models")
sys.path.append("./utils")

from typing import Optional
from binary_tree_node import TreeNode
from deserialize_to_binary_tree import DeserializeFromList
from serialize_binary_tree import SerializeBinaryTreeToList


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        """time complexity: O(n), this problem is to find out the "longest path"""
        #  NOTE: hint
        # |path rooted at node|= height(left)+height(right)
        # height(node)=1+max(height(left), height(right))
        # NOTE: This is a skill to use a list to pass by reference
        max_diameter: list[int] = [0]
        self.dfs(root, max_diameter)
        return max_diameter[0]

    def dfs(self, root: Optional[TreeNode], max_diameter: list[int]) -> int:
        """postorder traversal"""
        if root is None:
            return 0
        left: int = self.dfs(root.left, max_diameter)
        right: int = self.dfs(root.right, max_diameter)
        max_diameter[0] = max(max_diameter[0], left + right)
        return 1 + max(left, right)


def test_diameterOfBinaryTree_case_1():
    # arrange
    arr: list[int] = [1, 2, 3, 4, 5]
    root: Optional[TreeNode] = DeserializeFromList(arr)
    if root is None:
        raise Exception("failed")
    expected: int = 3

    # act
    solution = Solution()
    actual = solution.diameterOfBinaryTree(root)

    # assert
    assert expected == actual


def test_diameterOfBinaryTree_case_2():
    # arrange
    arr: list[int] = [1, 2]
    root: Optional[TreeNode] = DeserializeFromList(arr)
    if root is None:
        raise Exception("failed")
    expected: int = 1

    # act
    solution = Solution()
    actual = solution.diameterOfBinaryTree(root)

    # assert
    assert expected == actual


def test_diameterOfBinaryTree_case_3():
    # arrange
    arr: list[int] = [
        4,
        -7,
        -3,
        None,
        None,
        -9,
        -3,
        9,
        -7,
        -4,
        None,
        6,
        None,
        -6,
        -6,
        None,
        None,
        0,
        6,
        5,
        None,
        9,
        None,
        None,
        -1,
        -4,
        None,
        None,
        None,
        -2,
    ]
    root: Optional[TreeNode] = DeserializeFromList(arr)
    if root is None:
        raise Exception("failed")
    expected: int = 8

    # act
    solution = Solution()
    actual = solution.diameterOfBinaryTree(root)

    # assert
    assert expected == actual
